### DISCUSSION TITLE
Feedback and suggestions on how to improve scoreV2 in lazer

### DISCUSSION THREAD

**Discussion by rodrigopina360** [ppy/osu#24104] [Ideas]
Firstly sorry if this is not the correct way of sharing my feedback about the matter, and I hope that I'm not suggesting something that was already suggested in the past or that would be impossible to implement programatically.

# First combo break being too punishing
When spectating the score discussion meeting, one of the issues stated with scoreV2 was that the first couple misses were too punishing. I strongly agree that the first miss, especially when in the first 20-40% of the map is too punishing, and stagnate score too much.

# My suggestion
When you miss, use the maximum combo possible (after the miss) to scale combo, in order to not stagnate it when you miss, but only in the end of the map.

# Example
If you're playing a 2000 combo map, and miss on the first 500 combo, make it so the score increases as if you were fcing the last 1500 combo. If you then miss again 750 combo after that, scale the remaining score with the remaining 750 combo.
That change would make it so scoring would more rapidly increase in the start, and as you were playing the map and missing that score would start to increase slower and slower, pretty much stagnating in the end.

I believe a similar implementation could be done with acc, as currently dropping a couple 100s at the start of the map completely kills your score, only to go up as accuracy slowly climbed from that point.
Besides that, I strongly agree scoreV2 is the way to go, and that a 70/30 approach to combo/acc would work, but I personally would change it to 60/40, to give even more of an importance to acc, while combo being the main factor in scoring.

**Comment by monochrome22** [comment]
Correct me if I'm wrong, but wouldn't scaling the remaining score with the remaining combo mean that you will always get 100% of the score from combo?

**Reply by rodrigopina360** [reply]
Oh yeah, that was kind of a big oversight from my part. To remedy it my suggestion could be changed to scale with the remaining combo minus the already achieved combo, but that would have the side effect of rewarding no further combo score if you missed in the middle.
Some alternative could possibly be found to account smaller combos after the middle of the map, but currently I have no ideas how to do so, so I'll leave the discussion here and if I or anyone else has any ideas how to improve it they could add to this, but otherwise it's safe to say it could be closed.

**Comment by monochrome22** [comment]
Edit: this idea is already known. 

If we're throwing ideas around, for dealing with "a miss in the middle halving score", I had an idea to make it so you get the same amount of combo score from 100 combo as 1000 combo. This would make it more about punishing constant misses rather than one miss.

To explain it:
It could be max combo of the map / 10. So a map with 1000 combo, you would start gaining maximum score when you reach 100 combo. 
Going from 0 to 100 would progressively increase how much score you're getting, and once you reach 100, it would start giving maximum amount of score. So going from 0 to 100 would give you, say, 5% of the total score, and going from 100 to 200 would give you 10.555% of the score, and going from 800 to 900 would also give you 10.555% of the score. So going from 100 to 1000 would give you 95% of the score.

In practice, if there are two players, and one of them misses in the middle, it would mean they only get 95% of the total score from combo, while other player gets 100%. 
If one of the players misses every 100 combo, it would mean they only get 50% of the total score from combo (around 635000 score total).
Also, it would mean if one player has 800 + 100 + 100 combo, they would get 90% of the score, while a player with 500 + 500 combo would get 95% of the score. So a score with max combo 500 would beat score with max combo 800. This might look weird to players. A larger emphasis might need to be placed on showing the miss count to make this more presentable, or show how much combo was held in the maximum score combo range. Meaning, 800+100+100 would have 700 combo in the max combo range, while 500+500 would have 800 combo in the max combo range. I don't know if it's possible to have a solution for miss in the middle halving score while also having it so highest combo always has higher score. Though it can already happen that a score with 700 combo and constant misses loses to a score with 500 combo and one miss, so having a lower combo score be higher might not be a big deal.

If it's a problem that it's too easy to get a high score, then instead of max combo / 10, it could be max combo / 4 or something.

The downside of this idea is that it would make the scoring system more complicated and less predictable (which apparently developers don't like).

**Reply by rodrigopina360** [reply]
Yeah I think besides finding a good sweet spot for that max combo divider, it feels like a sound idea. I think a 6 or 7 would do the job, because when you account with the acc portion of the score it leads to what I believe to be a healthy score difference. 
A bit unrelated but is there a project or any information on how to simulate scores with scoreV2 as a line graph, or to export scores to a excel sheet like ppy showed in the meeting?

**Comment by monochrome22** [comment]
Edit: this idea is already known. 

Another idea is to add a negative multiplier to the combo score gain whenever you miss. `Score loss multiplier = 1 - current combo / max combo`. The multiplier will add to your existing multiplier every time you miss. So on a 2000 combo map, if you miss at 500 combo, you will have 25% negative multiplier, if you again get 500 combo and miss, then the multiplier becomes 50%.
Essentially, this system rewards consistency. If you keep missing, you will be spending time hitting circles with a higher negative multiplier. If you're holding combo, you're getting score without having a higher negative multiplier pulling it down. 

In practice, on a 2000 combo map, if you get 1000 combo, you will get 350000 score from combo, and then when you miss, you will get a 0.5 multiplier on your incoming score. So if you FC the rest and get 1000 combo again, you will get 175000 score for the second half. Currently in ScoreV2, missing in the middle will give you 675000 score total. With this idea, you will get 825000 if you miss in the middle.
Another example is, on a 2000 combo map, if a player misses every 100 combo, they would get 367500 score total from combo. If a player misses every 500 combo, they would get 437500 score total from combo.
Currently in ScoreV2, in a 100 combo map, if a player misses after 10 combo, then FCs the rest, they get around 860000 score. With this idea, it would be 937000. If they miss after 30 combo and FC the rest, currently they get around 720000 score, with this idea it would be 853000.

The problem with this idea might be the score progression. You will gain score faster at the start than at the end (if you're getting misses). Also, you would be gaining less score despite having the same combo. But the final score might be more acceptable for players. One option to help with confusion regarding same combo giving less score would be to display the score loss multiplier on screen. So if someone misses, they will be able to see that the multiplier is low, so it won't be confusing when they get less score despite having the same combo. And the player would know that the multiplier increases when you miss and get further into the map. 
Another problem is that every play will have relatively high score. You would have to be trying to get a score under 300000. 

I'm not good at math, so I can't write what the exact ScoreV2 formula would have to be to make this idea work.

**Reply by rodrigopina360** [reply]
I personally loved that idea, I don't even think high score would even be an issue, especially when you consider there are pass players/new players that would have very little combo/acc. It would be somewhat rewarding to pass, and would also make low miss count scores less punishable.

**Reply by monochrome22** [reply]
Apparently both of my ideas are already known, but I'll keep them in this thread so people can know that these solutions exist and discuss them if they want (and not make the same mistake of suggesting what's already known).

**Comment by monochrome22** [comment]
Another idea would be to use a multiplier based on current combo: `1.25 - (current combo / map max combo / 2)`.
The idea is that at smaller combos, you would get more score (than currently) from the multiplier, while at higher combos you would get less score (than currently). This would result in smaller combos mattering more, score not halving in the middle and few misses not mattering as much. The multiplier should work in a way that an FC adds to 700000, and a higher combo should still always give more score than a smaller combo. 
This idea is also flexible, because you can adjust the multiplier to a point that players are happy with, maybe gaining 75% of score from a miss in the middle would be good (compared to current 50%). 
It might be viable to add the multiplier to the judgements. 

This idea was also discussed in the [Discord](https://discord.com/channels/188630481301012481/1097318920991559880/1126212984579051622).

Edit: After further testing, it seems that the point where a higher combo starts giving less score than the previous combo would be reached quickly. With this idea, it's only possible to at most make a miss in the middle give 445000 combo score (63.5% of total combo score) instead of 356000 without diminishing combo score gain. The formula would be `2 - (current combo / map max combo)`.

If a high combo giving diminishing returns is not an issue and you would like to have around 75% of total score with a miss in the middle, then use `1.33 - (current combo / map max combo)`. With 1.33 multiplier, you will start gaining diminishing returns once you obtain 65% of map max combo.
Exactly how much score you gain from a miss in the middle also depends on the max combo of the map and the score is increased because of `/10d`, but 1.33 should be around 75% total combo score for most common combo ranges.
To be clear, your score would still be going up at high combos, you would just be gaining less score from each combo. 

To ensure the combo score doesn't go above 700000 at very low max map combos, it might be necessary to add `(current combo + 2)` and `(map max combo + 3)`. I used 2 and 3, because with my original formula (1.25 and /2), this caused the multipliers to become symmetrical, as in, 1.25, 1.15 1.05, 0.95, 0.85, 0.75, so this seems like the "proper" solution. Of course, you could just add map max combo + 2 and it will force the score below 700000.
If you want to test combo score going above 700000, then make a map with 9 notes and miss the one in the middle. At this point, the spike in score is the highest. 

For my testing, I added the formula to GetComboScoreChange: 
```
protected virtual double GetComboScoreChange(JudgementResult result) => Judgement.ToNumericResult(result.Type) * (1.33 - Convert.ToDouble(Combo.Value) / 643) * (1 + result.ComboAfterJudgement / 10d);

``` 
(I couldn't figure out how to make it use map max combo, so I just manually wrote the number and made all my maps 643 combo.)

Some numbers from a map with 643 combo (includes accuracy portion, all scores are 0.1x because I used Relax mod for testing):
- miss at 25% and FC the rest
2 - 80691
1.67 - 82905
1.5 - 84726
1.33 - 87474
default scorev2 - 73929

- miss every 10%x5 and FC the rest 50%
2 - 56328
1.67 - 58439
1.5 - 60176
1.33 - 62796
default scorev2 - 49881

- miss in middle
2 - 74514
1.67 - 77454
1.5 - 79872
1.33 - 83522 
default scorev2 - 65535

- miss every 25%
2 - 54724
1.67 - 57050
1.5 - 58963
1.33 - 61850
default scorev2 - 47622

- miss every 10%
2 - 38874
1.67 - 40202
1.5 - 41300
1.33 - 42965
default scorev2 - 34860

To be clear, the way it works is: normally you get +30 score with every subsequent combo. With this idea, with `2 -`, you will instead get +60 score on the first hit of a combo and +0 score on the last hit of a full combo, decreasing linearly relative to combo (at least that's what I think it does, I don't fully understand the code).
With `2 -`, it goes from `60 to 0`.
With `1.67 -`, it goes from `50 to -10`.
With `1.5 -`, it goes from `45 to -15`.
With `1.33 -`, it goes from `40 to -20`.

Another benefit of this idea is that it's very simple. I think anyone can understand the concept of the scoring being weighted towards lower combos, meaning 1% combo gives 99% additional score while 99% combo gives 1% additional score.

For making smaller combos matter more and increasing score for miss in the middle, one option would be to have two multipliers, one for first 50% of full combo and other for the second half. So going from 0% to 50% would have multiplier with score going from 60 to 20, then for second half it would use second multiplier of 20 to 0. 
However, this would make the code more complicated and the final result would probably be very similar to using square root of combo, so unless if there are problems with that idea, it can be used instead of mine.

**Comment by Zyfarok** [comment]
Before going any further, I suggest you give a look at the suggestion of square-root combo and provide some feedback on it first : https://github.com/ppy/osu/pull/24166
There's a spreadsheet along with the PR which should help you have a rough idea of what the final scores look like.
This solution is more in line with what has been done in other game-modes (aka rulesets), where they use a logarithm of the combo.
If you have any issue with that solution and you want to suggest something else, It would be nice if you could point out what those issues are and some spreadsheet comparing your solution to that one.

Here is some feedback on the above suggestions:

@Pinossaur as monochrome said I believe your suggestion doesn't work, and the other alternative you proposed doesn't solve the main problem of the miss in the middle.
One additional note : you said you'd prefer 60/40 to give more weight to accuracy, but there's other better ways to achieve this and in particular I believe that even a 99% acc score that has almost 0 combo shouldn't be given 40% of the score. If missing has a lower cost overall, then this will make accuracy much more important in comparison already, and if we want to give it more weight then we can also put an accuracy exponent on the combo portion of the score which is what I did in my suggestion. This latter points should also help in making accuracy still relevant even at low acc, compared to the current situation where a FC with 70% acc and a FC with 80% acc will give a similar score v2 (~500k versus ~590k).

@monochrome22 You suggest multiple things :
- Capping the combo at 10% of max_combo: this is very similar to what is done in taiko/catch/mania and I suggest you look at those (look at the classes overriding ScoreProcessor). The question is however: why would you cap at 10% and not any other value ? I believe this cap is not suited for very TV size maps. At the opposite, a fixed cap of 100 would not be suited for marathon maps. Also, the derivative is not smooth, or said differently I believe we should use a "soft" cap, like a square-root (as I suggested) or a logarithm (that is already used in taiko/mania/catch, except that we would remove the hard cap). However I would say the issues you mentioned are non-issues: Of course a player can have a better maximum combo and have a lower score and that is a desirable behavior ! We don't want the score to only depend on the biggest combo.
- From my understanding, the second suggestion suggests to have a multiplier that goes down at every miss. As you mentioned yourself this solution has many issues and works in a counter intuitive way. One additional issue you didn't mention is that the order of combo affects the score.
- I understand what you're trying to do, but your current solution has a severe problem : The point at which you gain the most score is not at max_combo but somewhere in between. Indeed, your formula of everything that multiplies the judgement result is basically a 2nd degree polynomial of combo (Combo.Value and ComboAfterJudgement are exactly the same), and the curve is not monotonic : https://www.wolframalpha.com/input?i=plot+%281.33+-+x+%2F+1000%29*%281+%2B+x+%2F+10%29+for+x+in+%5B0%2C+1000%5D
Thus I suggest using a concave and monotonic fonction of the combo instead, which is why I use square-root but feel free to suggest other alternative functions.

**Reply by monochrome22** [reply]
Thank you for the reply.

---

> Capping the combo at 10% of max_combo: this is very similar to what is done in taiko/catch/mania

Yes, I found this out after re-watching the community meeting (smoogi mentioned it). I agree with you that it's not good for osu!standard, as players are already used to a different progression system.

---

> From my understanding, the second suggestion suggests to have a multiplier that goes down at every miss. As you mentioned yourself this solution has many issues and works in a counter intuitive way. One additional issue you didn't mention is that the order of combo affects the score.

The decreasing multiplier idea is terrible as it breaks score progression, feel free to ignore it. I only left it up so other people know to not suggest it. Though I did point out the order of combo issue (if I understand correctly what you're referring to):
> You will gain score faster at the start than at the end (if you're getting misses). Also, you would be gaining less score despite having the same combo.
---


> I understand what you're trying to do, but your current solution has a severe problem : The point at which you gain the most score is not at max_combo but somewhere in between. 

I pointed this out multiple times in my post, I'm surprised you missed it:

> With this idea, it's only possible to at most make a miss in the middle give 445000 combo score (63.5% of total combo score) instead of 356000 without diminishing combo score gain. The formula would be `2 - (current combo / map max combo)`.
> With 1.33 multiplier, you will start gaining diminishing returns once you obtain 65% of map max combo.
> With 1.33 -, it goes from 40 to -20.

As I mentioned, a decent way to make my idea work would be to use 2 multipliers, one for first 50% combo and one for other 50% combo. For example, `2 - (combo/maxcombo / 0.75)`  for the first half and `1.33 - (combo/maxcombo / 1.5)` for the second half (or something like that). With the correct numbers, it might produce results very similar to the square root idea. Actually, with two multipliers it would be possible to make smaller combos matter even more than with square root, but I don't know if that's desirable. Using two multipliers ruins the simplicity of the idea though.

Anyway, the square root idea gives perfect final scores (in my opinion), has proper score progression and the implementation is more simple codewise, so almost certainly it will get used instead of mine. The only benefit of my idea is that the concept of judgements giving less score at higher combos due to a multiplier might be easier to understand than square root. For some it might be difficult to understand the implications of using square root.
Unless if there are some problems with the square root idea, it's not necessary to give my idea any further consideration (20 hours well spent).
